In [1]:
#importing necessery libraries for future analysis of the dataset
from IPython.display import display, HTML
import string

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import rcParams
import matplotlib.image as mpimg
from subprocess import check_output
from pprint import pprint
# from wordcloud import WordCloud, STOPWORDS

from collections import Counter

%matplotlib inline
pd.set_option("display.max_rows", None, "display.max_columns", None)

ModuleNotFoundError: No module named 'wordcloud'

In [ ]:
# from nltk.corpus import stopwords
# from nltk import WordNetLemmatizer
# from nltk import pos_tag,  sent_tokenize,word_tokenize, BigramAssocMeasures,\
#     BigramCollocationFinder, TrigramAssocMeasures, TrigramCollocationFinder
# print(check_output(["ls", "../input"]).decode("utf8"))

In [ ]:
#using pandas library and 'read_csv' function to read BlackFriday csv file as file already formated for us from Kaggle
airbnb = pd.read_csv("./airbnb/Airbnb_Listings.csv")

# examing head of BlackFriday csv file

airbnb.head(3)

id                         listing_url       scrape_id last_scraped  \
0   4952   https://www.airbnb.com/rooms/4952  20200530151957   2020-05-30   
1  11464  https://www.airbnb.com/rooms/11464  20200530151957   2020-05-31   
2  17884  https://www.airbnb.com/rooms/17884  20200530151957   2020-05-31   

                                name  \
0  Butterfly Inn - Graceful Living !   
1  Deluxe Private Studio-custom int.   
2               Silicon Valley Suite   

                                             summary  \
0  Lovely garden setting in a serene and art-fill...   
1  Custom built Studio with exquisite design. Rea...   
2  A guest suite for one or two, in a house in a ...   

                                               space  \
0  Very comfortable Queen bed and small desk in b...   
1  Description A favorite for international corpo...   
2  This is a private suite at the rear of a house...   

                                         description experiences_offered  \
0  Lovely garden setting in a serene and art-fill...                none   
1  Custom built Studio with exquisite design. Rea...                none   
2  A guest suite for one or two, in a house in a ...                none   

                               neighborhood_overview  \
0  Located in Professorville, Palo Alto, one of t...   
1  10 min. to Lucky's and Starbucks at El Camino ...   
2  This is a very quiet family neighborhood, but ...   

                                               notes  \
0  Housekeeping every Monday, leave your door ope...   
1  Pet Policy: - Well Behaved pet up to 25 lb. of...   
2  I cannot accommodate cats, sorry. I can usuall...   

                                             transit  \
0  Walking distance to Stanford University (30 mi...   
1  Public transportation at Homestead and Pomeroy...   
2  The CalTrain station is in walking distance, a...   

                                              access  \
0  Kitchen, laundry, family/TV room, garden, free...   
1          Complimentary Wifi-internet + Basic Cable   
2  Private outdoor patio. Shared washer/dryer on ...   

                                         interaction  \
0  Welcome and Orientation. Shared kitchen, garde...   
1  Available any time for support and problem res...   
2  I am usually in town.  Guests have a private e...   

                                         house_rules  thumbnail_url  \
0  No shoes in the house beyond the hall or laund...            NaN   
1  Can house a couple- bed type is a full size be...            NaN   
2  This is a quiet neighborhood, and the suite sh...            NaN   

   medium_url                                        picture_url  \
0         NaN  https://a0.muscache.com/im/pictures/529448/3d7...   
1         NaN  https://a0.muscache.com/im/pictures/121321/2c5...   
2         NaN  https://a0.muscache.com/im/pictures/08c0f0aa-3...   

   xl_picture_url  host_id                                 host_url host_name  \
0             NaN     7054   https://www.airbnb.com/users/show/7054   Mei Lin   
1             NaN    42458  https://www.airbnb.com/users/show/42458      Emey   
2             NaN    69023  https://www.airbnb.com/users/show/69023     Janet   

   host_since                             host_location  \
0  2009-01-24      Palo Alto, California, United States   
1  2009-09-30  Mountain View, California, United States   
2  2010-01-08        Lynn Haven, Florida, United States   

                                          host_about host_response_time  \
0  I have lived and worked in Asia, Europe, Austr...     within an hour   
1  I pride myself on attention to details and cus...     within an hour   
2  I am a software engineer, recently retired fro...       within a day   

  host_response_rate host_acceptance_rate host_is_superhost  \
0               100%                  98%                 t   
1               100%                  90%                 f   
2               100%                  50%     

In [ ]:
feature_df = airbnb[['id', 'listing_url', 'name']]


In [ ]:
feature_df.head()

id                         listing_url  \
0   4952   https://www.airbnb.com/rooms/4952   
1  11464  https://www.airbnb.com/rooms/11464   
2  17884  https://www.airbnb.com/rooms/17884   
3  21373  https://www.airbnb.com/rooms/21373   
4  37512  https://www.airbnb.com/rooms/37512   

                                  name  
0    Butterfly Inn - Graceful Living !  
1    Deluxe Private Studio-custom int.  
2                 Silicon Valley Suite  
3  Bonsai Garden Inn in Professorville  
4    Private  room - Parking 3 carport

In [ ]:
neighbourhoods = pd.read_csv("./airbnb/neighbourhoods.csv")
neighbourhoods.head(100)

neighbourhood_group         neighbourhood
0                   NaN              Campbell
1                   NaN             Cupertino
2                   NaN                Gilroy
3                   NaN             Los Altos
4                   NaN       Los Altos Hills
5                   NaN             Los Gatos
6                   NaN              Milpitas
7                   NaN          Monte Sereno
8                   NaN           Morgan Hill
9                   NaN         Mountain View
10                  NaN             Palo Alto
11                  NaN              San Jose
12                  NaN           Santa Clara
13                  NaN              Saratoga
14                  NaN             Sunnyvale
15                  NaN  Unincorporated Areas

In [ ]:
airbnb.describe(include='all')

id                        listing_url     scrape_id  \
count   7.221000e+03                               7221  7.221000e+03   
unique           NaN                               7221           NaN   
top              NaN  https://www.airbnb.com/rooms/4952           NaN   
freq             NaN                                  1           NaN   
mean    2.623922e+07                                NaN  2.020053e+13   
std     1.236420e+07                                NaN  1.781373e+00   
min     4.952000e+03                                NaN  2.020053e+13   
25%     1.624242e+07                                NaN  2.020053e+13   
50%     2.818432e+07                                NaN  2.020053e+13   
75%     3.749948e+07                                NaN  2.020053e+13   
max     4.359134e+07                                NaN  2.020053e+13   

       last_scraped                                              name  \
count          7221                                              7220   
unique            2                                              7077   
top      2020-05-31  TRIBE ❤️ Deep Cleaned  2BR @ Great Mall Milpitas   
freq           4782                                                21   
mean            NaN                                               NaN   
std             NaN                                               NaN   
min             NaN                                               NaN   
25%             NaN                                               NaN   
50%             NaN                                               NaN   
75%             NaN                                               NaN   
max             NaN                                               NaN   

                                                  summary  \
count                                                6922   
unique                                               6051   
top     The HackerHome Network New to the Valley? Tran...   
freq                                                   44   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                                    space  \
count                                                5514   
unique                                               4402   
top     At Zeus, we focus exclusively on 30 day+ stays...   
freq                                                  411   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                                              description experiences_offered  \
count                                                6983                7221   
unique                                               6439                   1   
top     Dorm-style house in Silicon Valley. No-host, 5...                none   
freq                                                   20                7221   
mean                                                  NaN                 NaN   
std                                                   NaN                 NaN   
min                                                   NaN                 NaN   
25%                                                   NaN                 NaN   
50%                                                   NaN            

In [ ]:
len(airbnb)

7221

In [ ]:
#checking type of every column in the dataset
airbnb.dtypes

id                                                int64
listing_url                                      object
scrape_id                                         int64
last_scraped                                     object
name                                             object
summary                                          object
space                                            object
description                                      object
experiences_offered                              object
neighborhood_overview                            object
notes                                            object
transit                                          object
access                                           object
interaction                                      object
house_rules                                      object
thumbnail_url                                   float64
medium_url                                      float64
picture_url                                     

In [ ]:
#.sort_values(ascending=True)
percent_missing = airbnb.isnull().sum() * 100 / len(airbnb)
missing_value_df = pd.DataFrame({
    'column_name': airbnb.columns,
    'percent_missing': percent_missing
}).sort_values(by='percent_missing', ascending=False)
missing_value_df

column_name  \
neighbourhood_group_cleansed                                  neighbourhood_group_cleansed   
thumbnail_url                                                                thumbnail_url   
license                                                                            license   
xl_picture_url                                                              xl_picture_url   
medium_url                                                                      medium_url   
square_feet                                                                    square_feet   
weekly_price                                                                  weekly_price   
monthly_price                                                                monthly_price   
notes                                                                                notes   
access                                                                              access   
transit                                                                            transit   
interaction                                                                    interaction   
neighborhood_overview                                                neighborhood_overview   
host_about                                                                      host_about   
house_rules                                                                    house_rules   
host_response_rate                                                      host_response_rate   
host_response_time                                                      host_response_time   
space                                                                                space   
security_deposit                                                          security_deposit   
review_scores_value                                                    review_scores_value   
review_scores_checkin                                                review_scores_checkin   
review_scores_location                                              review_scores_location   
review_scores_communication                                    review_scores_communication   
review_scores_cleanliness                                        review_scores_cleanliness   
review_scores_accuracy                                              review_scores_accuracy   
review_scores_rating                                                  review_scores_rating   
first_review                                                                  first_review   
last_review                                                                    last_review   
reviews_per_month                                                        reviews_per_month   
host_neighbourhood                                                      host_neighbourhood   
jurisdiction_names                                                      jurisdiction_names   
cleaning_fee                                                                  cleaning_fee   
host_acceptance_rate                                                  host_acceptance_rate   
neighbourhood                                                                neighbourhood   
summary                                                                            summary   
description                                                                    description   
beds                                                                                  beds   
zipcode                                                                            zipcode   
host_location                                                                host_location   
market                                                                              market   
bedrooms                                                                          bedrooms   
host_listings_count                                                    host_listings_count   
name                                 

In [ ]:
# find missing values and single value columns & url in data:
#[i for i in airbnb.columns if "url" in i]
del_cols = []

for col in airbnb.columns:
    if airbnb[col].isna().all():
        del_cols.append(col)
    elif airbnb[col].nunique() == 1:
        del_cols.append(col)
    elif "url" in col:
        del_cols.append(col)

del_cols

['listing_url',
 'scrape_id',
 'experiences_offered',
 'thumbnail_url',
 'medium_url',
 'picture_url',
 'xl_picture_url',
 'host_url',
 'host_thumbnail_url',
 'host_picture_url',
 'neighbourhood_group_cleansed',
 'country_code',
 'country',
 'has_availability',
 'requires_license',
 'license',
 'is_business_travel_ready']

In [ ]:
airbnb = airbnb.drop(del_cols, axis=1)


In [ ]:
#airbnb.columns
cat_features = airbnb.select_dtypes(include=['object']).columns
cat_features

Index(['last_scraped', 'name', 'summary', 'space', 'description',
       'neighborhood_overview', 'notes', 'transit', 'access', 'interaction',
       'house_rules', 'host_name', 'host_since', 'host_location', 'host_about',
       'host_response_time', 'host_response_rate', 'host_acceptance_rate',
       'host_is_superhost', 'host_neighbourhood', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'street',
       'neighbourhood', 'neighbourhood_cleansed', 'city', 'state', 'zipcode',
       'market', 'smart_location', 'is_location_exact', 'property_type',
       'room_type', 'bed_type', 'amenities', 'price', 'weekly_price',
       'monthly_price', 'security_deposit', 'cleaning_fee', 'extra_people',
       'calendar_updated', 'calendar_last_scraped', 'first_review',
       'last_review', 'jurisdiction_names', 'instant_bookable',
       'cancellation_policy', 'require_guest_profile_picture',
       'require_guest_phone_verification'],
      dtype='object')

In [ ]:
cat_cols = []

for col in cat_features:
    list_unique = (col, len(airbnb[col].unique()))
    print(col)
    print(airbnb[col].value_counts(normalize=True).loc[lambda x: x > 0.2])
    cat_cols.append(list_unique)
    #print(list_unique)

#output = sorted(cat_cols, key=lambda x: x[-1])
#output

last_scraped
2020-05-31    0.662235
2020-05-30    0.337765
Name: last_scraped, dtype: float64
name
Series([], Name: name, dtype: float64)
summary
Series([], Name: summary, dtype: float64)
space
Series([], Name: space, dtype: float64)
description
Series([], Name: description, dtype: float64)
neighborhood_overview
Series([], Name: neighborhood_overview, dtype: float64)
notes
Series([], Name: notes, dtype: float64)
transit
Series([], Name: transit, dtype: float64)
access
Series([], Name: access, dtype: float64)
interaction
Series([], Name: interaction, dtype: float64)
house_rules
Series([], Name: house_rules, dtype: float64)
host_name
Series([], Name: host_name, dtype: float64)
host_since
Series([], Name: host_since, dtype: float64)
host_location
San Jose, California, United States    0.268964
Name: host_location, dtype: float64
host_about
Series([], Name: host_about, dtype: float64)
host_response_time
within an hour    0.737532
Name: host_response_time, dtype: float64
host_response_rate


host_response_time
within an hour 0.737532
property_type
House 0.514472

host_response_rate
100% 0.810676

host_acceptance_rate
100% 0.340864

host_is_superhost
f 0.552355
t 0.447645

host_identity_verified
f 0.562465
t 0.437535

room_type
Private room 0.478327
Entire home/apt 0.475142

extra_people
$0.00 0.472788

cancellation_policy
flexible 0.390112
moderate 0.315469
strict_14_with_grace_period 0.294281

instant_bookable
f 0.55214
t 0.44786

require_guest_profile_picture
f 0.968148

require_guest_phone_verification
f 0.968425


In [ ]:
feature_df['host_response_in_1hr'] = airbnb['host_response_time'].apply(
    lambda x: int(x == 'within an hour') if isinstance(x, str) else 0)

    
feature_df['host_response_rate100'] = airbnb['host_response_rate'].apply(
    lambda x: int(x == '100%') if isinstance(x, str) else 0)
feature_df['host_acceptance_rate100'] = airbnb['host_acceptance_rate'].apply(
    lambda x: int(x == '100%') if isinstance(x, str) else 0)
feature_df['host_is_superhost'] = airbnb['host_is_superhost'].apply(
    lambda x: int(x == 't') if isinstance(x, str) else 0)
feature_df['host_identity_verified'] = airbnb['host_identity_verified'].apply(
    lambda x: int(x == 't') if isinstance(x, str) else 0)
feature_df['room_type_private_room'] = airbnb['room_type'].apply(
    lambda x: int(x == 'Private room') if isinstance(x, str) else 0)
feature_df['room_type_entire_home'] = airbnb['room_type'].apply(
    lambda x: int(x == 'Entire home/apt') if isinstance(x, str) else 0)
feature_df['charge_extra_people_fee'] = airbnb['extra_people'].apply(
    lambda x: int(x != '$0.00') if isinstance(x, str) else 0)

feature_df['cancellation_policy_flexible'] = airbnb[
    'cancellation_policy'].apply(lambda x: int(x == 'flexible')
                                 if isinstance(x, str) else 0)
feature_df['cancellation_policy_moderate'] = airbnb[
    'cancellation_policy'].apply(lambda x: int(x == 'moderate')
                                 if isinstance(x, str) else 0)
feature_df['cancellation_policy_strict'] = airbnb['cancellation_policy'].apply(
    lambda x: int(x == 'strict_14_with_grace_period')
    if isinstance(x, str) else 0)
feature_df['instant_bookable'] = airbnb['instant_bookable'].apply(
    lambda x: int(x == 't') if isinstance(x, str) else 0)


In [ ]:
feature_df.head()

id                         listing_url  \
0   4952   https://www.airbnb.com/rooms/4952   
1  11464  https://www.airbnb.com/rooms/11464   
2  17884  https://www.airbnb.com/rooms/17884   
3  21373  https://www.airbnb.com/rooms/21373   
4  37512  https://www.airbnb.com/rooms/37512   

                                  name  host_response_in_1hr  \
0    Butterfly Inn - Graceful Living !                     1   
1    Deluxe Private Studio-custom int.                     1   
2                 Silicon Valley Suite                     0   
3  Bonsai Garden Inn in Professorville                     1   
4    Private  room - Parking 3 carport                     0   

   host_response_rate100  host_acceptance_rate100  host_is_superhost  \
0                      1                        0                  1   
1                      1                        0                  0   
2                      1                        0                  0   
3                      1                        0                  1   
4                      1                        1                  1   

   host_identity_verified  room_type_private_room  room_type_entire_home  \
0                       0                       1                      0   
1                       0                       0                      1   
2                       1                       0                      1   
3                       0                       1                      0   
4                       1                       1                      0   

   charge_extra_people_fee  cancellation_policy_flexible  \
0                        1                             0   
1                        1                             0   
2                        0                             0   
3                        1                             0   
4                        0                             0   

   cancellation_policy_moderate  cancellation_policy_strict  instant_bookable  
0                             1                           0                 0  
1                             1                           0                 0  
2                             0                           1                 0  
3                             1                           0                 0  
4                             1                           0                 0

In [ ]:
numeric_features = airbnb.select_dtypes(include=['int64', 'float64']).columns
airbnb[numeric_features].describe()

id       host_id  host_listings_count  \
count  7.221000e+03  7.221000e+03          7220.000000   
mean   2.623922e+07  8.886443e+07           139.588504   
std    1.236420e+07  9.390593e+07           472.674992   
min    4.952000e+03  7.054000e+03             0.000000   
25%    1.624242e+07  1.697278e+07             1.000000   
50%    2.818432e+07  4.800549e+07             3.000000   
75%    3.749948e+07  1.429331e+08            10.000000   
max    4.359134e+07  3.477992e+08          2007.000000   

       host_total_listings_count     latitude    longitude  accommodates  \
count                7220.000000  7221.000000  7221.000000   7221.000000   
mean                  139.588504    37.352285  -121.966600      3.114804   
std                   472.674992     0.064819     0.108587      2.370579   
min                     0.000000    36.965600  -122.190350      1.000000   
25%                     1.000000    37.319770  -122.042450      2.000000   
50%                     3.000000    37.358260  -121.961940      2.000000   
75%                    10.000000    37.398250  -121.886340      4.000000   
max                  2007.000000    37.462980  -121.380120     16.000000   

         bathrooms     bedrooms         beds  square_feet  guests_included  \
count  7220.000000  7215.000000  7153.000000    29.000000      7221.000000   
mean      1.383449     1.421760     1.817000   691.000000         1.699349   
std       0.688715     0.999833     1.633911   735.833492         1.620945   
min       0.000000     0.000000     0.000000    50.000000         1.000000   
25%       1.000000     1.000000     1.000000   225.000000         1.000000   
50%       1.000000     1.000000     1.000000   396.000000         1.000000   
75%       2.000000     2.000000     2.000000   950.000000         2.000000   
max       8.000000     9.000000    54.000000  3600.000000        16.000000   

       minimum_nights  maximum_nights  minimum_minimum_nights  \
count     7221.000000    7.221000e+03             7221.000000   
mean         9.756959    2.980880e+05                9.142916   
std         34.689846    2.527150e+07               35.027086   
min          1.000000    1.000000e+00                1.000000   
25%          1.000000    6.000000e+01                1.000000   
50%          2.000000    1.125000e+03                2.000000   
75%          7.000000    1.125000e+03                5.000000   
max       1125.000000    2.147484e+09             1125.000000   

       maximum_minimum_nights  minimum_maximum_nights  maximum_maximum_nights  \
count             7221.000000            7.221000e+03            7.221000e+03   
mean                14.247888            2.982325e+05            1.785205e+06   
std                 63.358607            2.527150e+07            6.188084e+07   
min                  1.000000            1.000000e+00            1.000000e+00   
25%                  1.000000            3.650000e+02            3.650000e+02   
50%                  2.000000            1.125000e+03            1.125000e+03   
75%                  7.000000            1.125000e+03            1.125000e+03   
max               1125.000000            2.147484e+09            2.147484e+09   

       minimum_nights_avg_ntm  maximum_nights_avg_ntm  availability_30  \
count             7221.000000            7.221000e+03      7221.000000   
mean                13.631505            1.538945e+06        13.689101   
std                 61.691820            5.347742e+07        13.616884   
min                  1.000000            1.000000e+00         0.000000   
25%                  1.000000            3.650000e+02         0.000000   
50%                  2.000000            1.125000e+03        10.000000   
75%                  7.000000            1.125000e+03        29.000000   
max               1125.000000            2.147484e+09        30.000000   

       availability_60  availability_90  availability_365  number_of_reviews  \
count      7221.000000      7221.000000    

# Clean price


In [ ]:
## apply function to multiple columns
def price_cov(x):
    return x.str.replace('$', '').str.replace(',', '').astype(float)

price_colname = [
    'price', 'security_deposit', 'cleaning_fee', 'monthly_price',
    'weekly_price'
]
airbnb[price_colname] = airbnb[price_colname].apply(price_cov, axis=1)


/var/folders/dv/qnpc2tm905n3c2nrmq03c65w0000gn/T/ipykernel_47751/1405469132.py:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  return x.str.replace('$', '').str.replace(',', '').astype(float)


In [ ]:
airbnb[price_colname].head()

price  security_deposit  cleaning_fee  monthly_price  weekly_price
0   90.0               0.0          35.0         1428.0         440.0
1   85.0             700.0         250.0         2100.0           NaN
2   65.0            1000.0           NaN         2800.0        1400.0
3   99.0               0.0          35.0         2000.0         600.0
4   52.0               NaN           NaN         1386.0         393.0

In [ ]:
## missing value inputation
## monthly_price weekly_price
price_complete = airbnb[airbnb['monthly_price'].notnull()
                        & airbnb['weekly_price'].notnull()][[
                            'monthly_price', 'weekly_price', 'price',
                            'cleaning_fee', 'security_deposit'
                        ]]
price_complete.sort_values(by='price')
# create columes weekly_special_price monthly_special_price

price_complete['weekly_special_price'] = price_complete[
    'price'] * 7 - price_complete['weekly_price']
price_complete['monthly_special_price'] = price_complete[
    'price'] * 31 - price_complete['monthly_price']

price_complete.fillna(0, inplace=True)

In [ ]:
price_complete.head(10)

monthly_price  weekly_price  price  cleaning_fee  security_deposit  \
0          1428.0         440.0   90.0          35.0               0.0   
2          2800.0        1400.0   65.0           0.0            1000.0   
3          2000.0         600.0   99.0          35.0               0.0   
4          1386.0         393.0   52.0           0.0               0.0   
5          2400.0         875.0  110.0          35.0             100.0   
6          1450.0         450.0   60.0          20.0             100.0   
7          2200.0         800.0  125.0          35.0             100.0   
9          1950.0         700.0  115.0          35.0               0.0   
10         2150.0         675.0   95.0          30.0               0.0   
11         3699.0        1099.0  189.0           0.0             500.0   

    weekly_special_price  monthly_special_price  
0                  190.0                 1362.0  
2                 -945.0                 -785.0  
3                   93.0                 1069.0  
4                  -29.0                  226.0  
5                 -105.0                 1010.0  
6                  -30.0                  410.0  
7                   75.0                 1675.0  
9                  105.0                 1615.0  
10                 -10.0                  795.0  
11                 224.0                 2160.0

In [ ]:
## append selected price_colname to feature_df
feature_df = pd.concat([
    feature_df, airbnb[['price', 'security_deposit', 'cleaning_fee']].fillna(0)
],
                       axis=1)


In [ ]:
feature_df = pd.concat([
    feature_df,
    price_complete[['weekly_special_price', 'monthly_special_price']]
],
                       axis=1).fillna(0)
feature_df.head()

id                         listing_url  \
0   4952   https://www.airbnb.com/rooms/4952   
1  11464  https://www.airbnb.com/rooms/11464   
2  17884  https://www.airbnb.com/rooms/17884   
3  21373  https://www.airbnb.com/rooms/21373   
4  37512  https://www.airbnb.com/rooms/37512   

                                  name  host_response_in_1hr  \
0    Butterfly Inn - Graceful Living !                     1   
1    Deluxe Private Studio-custom int.                     1   
2                 Silicon Valley Suite                     0   
3  Bonsai Garden Inn in Professorville                     1   
4    Private  room - Parking 3 carport                     0   

   host_response_rate100  host_acceptance_rate100  host_is_superhost  \
0                      1                        0                  1   
1                      1                        0                  0   
2                      1                        0                  0   
3                      1                        0                  1   
4                      1                        1                  1   

   host_identity_verified  room_type_private_room  room_type_entire_home  \
0                       0                       1                      0   
1                       0                       0                      1   
2                       1                       0                      1   
3                       0                       1                      0   
4                       1                       1                      0   

   charge_extra_people_fee  cancellation_policy_flexible  \
0                        1                             0   
1                        1                             0   
2                        0                             0   
3                        1                             0   
4                        0                             0   

   cancellation_policy_moderate  cancellation_policy_strict  instant_bookable  \
0                             1                           0                 0   
1                             1                           0                 0   
2                             0                           1                 0   
3                             1                           0                 0   
4                             1                           0                 0   

   price  security_deposit  cleaning_fee  weekly_special_price  \
0   90.0               0.0          35.0                 190.0   
1   85.0             700.0         250.0                   0.0   
2   65.0            1000.0           0.0                -945.0   
3   99.0               0.0          35.0                  93.0   
4   52.0               0.0           0.0                 -29.0   

   monthly_special_price  
0                 1362.0  
1                    0.0  
2                 -785.0  
3                 1069.0  
4                  226.0

In [ ]:
len(feature_df)

7221

# Clean location


In [ ]:
airbnb[[
    'street', 'neighbourhood', 'neighbourhood_cleansed', 'city', 'state',
    'zipcode', 'market', 'smart_location', 'latitude', 'longitude',
    'is_location_exact'
]].head(10)


street   neighbourhood neighbourhood_cleansed  \
0      Palo Alto, CA, United States       Palo Alto              Palo Alto   
1    Santa Clara, CA, United States     Santa Clara            Santa Clara   
2  Mountain View, CA, United States   Mountain View          Mountain View   
3      Palo Alto, CA, United States       Palo Alto              Palo Alto   
4       San Jose, CA, United States  South San Jose               San Jose   
5      Palo Alto, CA, United States       Palo Alto              Palo Alto   
6       San Jose, CA, United States     Santa Clara               San Jose   
7      Palo Alto, CA, United States       Palo Alto              Palo Alto   
8       San Jose, CA, United States  South San Jose               San Jose   
9      Palo Alto, CA, United States       Palo Alto              Palo Alto   

            city state zipcode         market     smart_location  latitude  \
0      Palo Alto    CA   94301  South Bay, CA      Palo Alto, CA  37.43972   
1    Santa Clara    CA   95051  South Bay, CA    Santa Clara, CA  37.34342   
2  Mountain View    CA   94041  South Bay, CA  Mountain View, CA  37.38612   
3      Palo Alto    CA   94301  South Bay, CA      Palo Alto, CA  37.44001   
4       San Jose    CA   95111  South Bay, CA       San Jose, CA  37.30598   
5      Palo Alto    CA   94301  South Bay, CA      Palo Alto, CA  37.45356   
6       San Jose    CA   95117  South Bay, CA       San Jose, CA  37.32654   
7      Palo Alto    CA   94301  South Bay, CA      Palo Alto, CA  37.45348   
8       San Jose    CA   95111  South Bay, CA       San Jose, CA  37.30521   
9      Palo Alto    CA   94301  South Bay, CA      Palo Alto, CA  37.43925   

   longitude is_location_exact  
0 -122.15672                 t  
1 -121.98829                 t  
2 -122.07357                 t  
3 -122.15667                 t  
4 -121.84036                 t  
5 -122.13733                 t  
6 -121.94955                 t  
7 -122.13714                 t  
8 -121.84149                 t  
9 -122.15617                 t

In [ ]:
np.sort(airbnb.neighbourhood_cleansed.unique()).tolist()

['Campbell',
 'Cupertino',
 'Gilroy',
 'Los Altos',
 'Los Altos Hills',
 'Los Gatos',
 'Milpitas',
 'Monte Sereno',
 'Morgan Hill',
 'Mountain View',
 'Palo Alto',
 'San Jose',
 'Santa Clara',
 'Saratoga',
 'Sunnyvale',
 'Unincorporated Areas']

In [ ]:
airbnb.market.unique().tolist()

['South Bay, CA', 'Monterey Region', 'Other (Domestic)', 'South Bay CA', nan]

In [ ]:
neighbourhoods.neighbourhood.tolist()

['Campbell',
 'Cupertino',
 'Gilroy',
 'Los Altos',
 'Los Altos Hills',
 'Los Gatos',
 'Milpitas',
 'Monte Sereno',
 'Morgan Hill',
 'Mountain View',
 'Palo Alto',
 'San Jose',
 'Santa Clara',
 'Saratoga',
 'Sunnyvale',
 'Unincorporated Areas']

In [ ]:
feature_df['neighbourhood_cleansed'] = airbnb.neighbourhood_cleansed
feature_df.head()

id                         listing_url  \
0   4952   https://www.airbnb.com/rooms/4952   
1  11464  https://www.airbnb.com/rooms/11464   
2  17884  https://www.airbnb.com/rooms/17884   
3  21373  https://www.airbnb.com/rooms/21373   
4  37512  https://www.airbnb.com/rooms/37512   

                                  name  host_response_in_1hr  \
0    Butterfly Inn - Graceful Living !                     1   
1    Deluxe Private Studio-custom int.                     1   
2                 Silicon Valley Suite                     0   
3  Bonsai Garden Inn in Professorville                     1   
4    Private  room - Parking 3 carport                     0   

   host_response_rate100  host_acceptance_rate100  host_is_superhost  \
0                      1                        0                  1   
1                      1                        0                  0   
2                      1                        0                  0   
3                      1                        0                  1   
4                      1                        1                  1   

   host_identity_verified  room_type_private_room  room_type_entire_home  \
0                       0                       1                      0   
1                       0                       0                      1   
2                       1                       0                      1   
3                       0                       1                      0   
4                       1                       1                      0   

   charge_extra_people_fee  cancellation_policy_flexible  \
0                        1                             0   
1                        1                             0   
2                        0                             0   
3                        1                             0   
4                        0                             0   

   cancellation_policy_moderate  cancellation_policy_strict  instant_bookable  \
0                             1                           0                 0   
1                             1                           0                 0   
2                             0                           1                 0   
3                             1                           0                 0   
4                             1                           0                 0   

   price  security_deposit  cleaning_fee  weekly_special_price  \
0   90.0               0.0          35.0                 190.0   
1   85.0             700.0         250.0                   0.0   
2   65.0            1000.0           0.0                -945.0   
3   99.0               0.0          35.0                  93.0   
4   52.0               0.0           0.0                 -29.0   

   monthly_special_price neighbourhood_cleansed  
0                 1362.0              Palo Alto  
1                    0.0            Santa Clara  
2                 -785.0          Mountain View  
3                 1069.0              Palo Alto  
4                  226.0               San Jose

# Clean room type


In [ ]:
room_colname = [
    'property_type', 'room_type', 'accommodates', 'bathrooms', 'bedrooms',
    'beds', 'bed_type', 'amenities', 'square_feet'
]
airbnb[room_colname].head(10)

property_type        room_type  accommodates  bathrooms  bedrooms  beds  \
0         Villa     Private room             2        2.0       1.0   1.0   
1     Apartment  Entire home/apt             1        1.0       NaN   1.0   
2   Guest suite  Entire home/apt             2        1.0       1.0   1.0   
3         Villa     Private room             1        2.0       1.0   1.0   
4      Bungalow     Private room             2        1.0       1.0   1.0   
5         House     Private room             2        1.0       1.0   1.0   
6         House     Private room             1        1.0       1.0   1.0   
7         House     Private room             2        1.0       1.0   1.0   
8      Bungalow     Private room             2        1.0       1.0   1.0   
9         Villa     Private room             2        2.0       1.0   4.0   

   bed_type                                          amenities  square_feet  
0  Real Bed  {TV,"Cable TV",Internet,Wifi,Kitchen,Heating,W...          NaN  
1  Real Bed  {TV,"Cable TV",Internet,Wifi,"Air conditioning...        325.0  
2  Real Bed  {TV,"Cable TV",Internet,Wifi,"Air conditioning...          NaN  
3  Real Bed  {TV,"Cable TV",Internet,Wifi,Kitchen,"Paid par...          NaN  
4  Real Bed  {Internet,Wifi,"Air conditioning",Kitchen,"Fre...          NaN  
5  Real Bed  {TV,"Cable TV",Internet,Wifi,"Pets live on thi...          NaN  
6  Real Bed  {TV,Wifi,"Free parking on premises","Pets live...          NaN  
7  Real Bed  {Internet,Wifi,"Free parking on premises","Pet...          NaN  
8     Futon  {Internet,Wifi,"Air conditioning",Kitchen,"Fre...          NaN  
9  Real Bed  {TV,"Cable TV",Internet,Wifi,Kitchen,"Indoor f...        100.0

In [ ]:
numeric_features = airbnb.select_dtypes(
    include=['int64', 'float64', 'float32', 'int32'])
nominal_features = airbnb.select_dtypes(include=['object'])

for item in room_colname:
    print(item)
    print(airbnb[item].unique())

property_type
['Villa' 'Apartment' 'Guest suite' 'Bungalow' 'House' 'Guesthouse' 'Loft'
 'Other' 'Condominium' 'Townhouse' 'Cottage' 'Bed and breakfast' 'Cabin'
 'Camper/RV' 'Tiny house' 'Serviced apartment' 'Treehouse' 'Tent' 'Train'
 'Barn' 'Yurt' 'Boutique hotel' 'Lighthouse' 'Farm stay' 'Campsite'
 'Earth house' 'Aparthotel' 'Chalet']
room_type
['Private room' 'Entire home/apt' 'Shared room' 'Hotel room']
accommodates
[ 2  1  4  3 16  6 12 10  8  5  7 14 13  9 15 11]
bathrooms
[2.  1.  0.5 1.5 3.  0.  2.5 4.5 nan 3.5 4.  6.  5.5 5.  6.5 8. ]
bedrooms
[ 1. nan  2.  5.  3.  8.  4.  0.  7.  6.  9.]
beds
[ 1.  4.  2.  3. 11.  8.  7.  0. 12.  5.  6. nan 10.  9. 17. 26. 13. 15.
 16. 20. 54.]
bed_type
['Real Bed' 'Futon' 'Pull-out Sofa' 'Airbed' 'Couch']
amenities
['{TV,"Cable TV",Internet,Wifi,Kitchen,Heating,Washer,Dryer,"Smoke alarm","Carbon monoxide alarm","First aid kit","Safety card","Fire extinguisher",Essentials,Shampoo,"Lock on bedroom door",Hangers,"Hair dryer",Iron,"Laptop-frie

In [ ]:
nominal_features = airbnb[room_colname].select_dtypes(include=['object'])

In [ ]:
nominal_features.head()

property_type        room_type  bed_type  \
0         Villa     Private room  Real Bed   
1     Apartment  Entire home/apt  Real Bed   
2   Guest suite  Entire home/apt  Real Bed   
3         Villa     Private room  Real Bed   
4      Bungalow     Private room  Real Bed   

                                           amenities  
0  {TV,"Cable TV",Internet,Wifi,Kitchen,Heating,W...  
1  {TV,"Cable TV",Internet,Wifi,"Air conditioning...  
2  {TV,"Cable TV",Internet,Wifi,"Air conditioning...  
3  {TV,"Cable TV",Internet,Wifi,Kitchen,"Paid par...  
4  {Internet,Wifi,"Air conditioning",Kitchen,"Fre...

In [ ]:
# Cat_features.head()
# pd.get_dummies(data,
#                       columns=[
#                           'host_is_superhost', 'host_identity_verified',
#                           'instant_bookable', 'cancellation_policy',
#                           'room_type'
#                       ],
#                       prefix=[
#                           'Superhost', 'Host verified', 'Instantbook',
#                           'Cancellation', 'Room type'
#                       ],
#                       dtype='int64')

In [ ]:
#transform into lower case and export
airbnb['amenities'].str.lower().to_csv("./airbnb/amenities.csv")

# Create Dummy Variable for top 50 amenities


[Where Condition Reference](https://towardsdatascience.com/3-methods-to-create-conditional-columns-with-python-pandas-and-numpy-a6cd4be9da53)


In [ ]:
results = Counter()
amenities = airbnb['amenities'].str.lower()
amenities.str.strip('{}')\
               .str.replace('"', '')\
               .str.lstrip('\"')\
               .str.rstrip('\"')\
               .str.split(',')\
               .apply(results.update)
sub_df = pd.DataFrame(results.most_common(50), columns=['amenities', 'count'])
sub_df['ratio'] = round(sub_df['count'] / len(amenities), 4)
sub_df.sort_values(by=['count'], ascending=False)


amenities  count   ratio
0                         wifi   7071  0.9792
1                   essentials   6904  0.9561
2                      heating   6824  0.9450
3                  smoke alarm   6794  0.9409
4                      hangers   6480  0.8974
5                      shampoo   6271  0.8684
6        carbon monoxide alarm   6178  0.8556
7    laptop-friendly workspace   5998  0.8306
8                   hair dryer   5997  0.8305
9                      kitchen   5890  0.8157
10                      washer   5817  0.8056
11                       dryer   5785  0.8011
12                        iron   5711  0.7909
13                   hot water   5280  0.7312
14    free parking on premises   5277  0.7308
15                          tv   5203  0.7205
16            air conditioning   4565  0.6322
17                refrigerator   4561  0.6316
18                   microwave   4541  0.6289
19       dishes and silverware   4139  0.5732
20         free street parking   4043  0.5599
21           fire extinguisher   4040  0.5595
22                coffee maker   3907  0.5411
23                  bed linens   3710  0.5138
24        lock on bedroom door   3677  0.5092
25               first aid kit   3668  0.5080
26                       stove   3455  0.4785
27              cooking basics   3447  0.4774
28                        oven   3434  0.4756
29               self check-in   3257  0.4510
30            private entrance   3052  0.4227
31  extra pillows and blankets   3024  0.4188
32     long term stays allowed   2851  0.3948
33                  dishwasher   2829  0.3918
34          garden or backyard   2437  0.3375
35            patio or balcony   2291  0.3173
36                    cable tv   2228  0.3085
37     luggage dropoff allowed   1839  0.2547
38                    internet   1838  0.2545
39           single level home   1512  0.2094
40                        pool   1437  0.1990
41         family/kid friendly   1429  0.1979
42            indoor fireplace   1345  0.1863
43                      keypad   1286  0.1781
44         private living room   1203  0.1666
45                     lockbox   1197  0.1658
46                     bathtub   1070  0.1482
47                pets allowed   1040  0.1440
48                   bbq grill   1028  0.1424
49                         gym   1008  0.1396

In [ ]:
airbnb['amenities'] = airbnb['amenities'].str.lower()
for item in sub_df.amenities.tolist():
    feature_df[item.replace(" ", "_")] = airbnb['amenities'].apply(
        lambda x: int(item in x) if isinstance(x, str) else 0)


In [ ]:
#airbnb['amenities']= airbnb['amenities'].str.lower()
#type(airbnb['amenities'][1])
#feature_df["wifi"] = np.where("wifi" in airbnb['amenities'],1, 0)

feature_df["wifi"].sum()
feature_df["wifi"].describe()

count    7221.000000
mean        0.979366
std         0.142166
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: wifi, dtype: float64

In [ ]:
# 'tv' in airbnb['amenities'].str.lower()[1]
# for item in sub_df.amenities.tolist():
#     print (item)
#     print (item in airbnb['amenities'][1])

In [ ]:
feature_df.describe()

id  host_response_in_1hr  host_response_rate100  \
count  7.221000e+03           7221.000000            7221.000000   
mean   2.623922e+07              0.552971               0.607811   
std    1.236420e+07              0.497221               0.488272   
min    4.952000e+03              0.000000               0.000000   
25%    1.624242e+07              0.000000               0.000000   
50%    2.818432e+07              1.000000               1.000000   
75%    3.749948e+07              1.000000               1.000000   
max    4.359134e+07              1.000000               1.000000   

       host_acceptance_rate100  host_is_superhost  host_identity_verified  \
count              7221.000000        7221.000000             7221.000000   
mean                  0.307160           0.447583                0.437474   
std                   0.461348           0.497279                0.496109   
min                   0.000000           0.000000                0.000000   
25%                   0.000000           0.000000                0.000000   
50%                   0.000000           0.000000                0.000000   
75%                   1.000000           1.000000                1.000000   
max                   1.000000           1.000000                1.000000   

       room_type_private_room  room_type_entire_home  charge_extra_people_fee  \
count             7221.000000            7221.000000              7221.000000   
mean                 0.478327               0.475142                 0.527212   
std                  0.499565               0.499416                 0.499294   
min                  0.000000               0.000000                 0.000000   
25%                  0.000000               0.000000                 0.000000   
50%                  0.000000               0.000000                 1.000000   
75%                  1.000000               1.000000                 1.000000   
max                  1.000000               1.000000                 1.000000   

       cancellation_policy_flexible  cancellation_policy_moderate  \
count                   7221.000000                   7221.000000   
mean                       0.390112                      0.315469   
std                        0.487809                      0.464734   
min                        0.000000                      0.000000   
25%                        0.000000                      0.000000   
50%                        0.000000                      0.000000   
75%                        1.000000                      1.000000   
max                        1.000000                      1.000000   

       cancellation_policy_strict  instant_bookable         price  \
count                 7221.000000       7221.000000   7221.000000   
mean                     0.294281          0.447860    161.287079   
std                      0.455750          0.497308    352.069000   
min                      0.000000          0.000000     10.000000   
25%                      0.000000          0.000000     60.000000   
50%                      0.000000          0.000000    100.000000   
75%                      1.000000          1.000000    175.000000   
max                      1.000000          1.000000  10000.000000   

       security_deposit  cleaning_fee  weekly_special_price  \
count       7221.000000   7221.000000           7221.000000   
mean         253.718460     84.544800              4.960809   
std          506.534144    106.107352            196.287203   
min            0.000000      0.000000          -5390.000000   
25%            0.000000     20.000000              0.000000   
50%          100.000000     50.000000              0.000000   
75%          350.000000    100.000000              0.000000   
max         5000.000000   1000.000000           6615.000000   

       monthly_special_price         wifi   essentials      heating  \
count            7221.000000  7221.000000  7221.000000  7221.000000   
mean               87.3984

In [ ]:
feature_df.head()

id                         listing_url  \
0   4952   https://www.airbnb.com/rooms/4952   
1  11464  https://www.airbnb.com/rooms/11464   
2  17884  https://www.airbnb.com/rooms/17884   
3  21373  https://www.airbnb.com/rooms/21373   
4  37512  https://www.airbnb.com/rooms/37512   

                                  name  host_response_in_1hr  \
0    Butterfly Inn - Graceful Living !                     1   
1    Deluxe Private Studio-custom int.                     1   
2                 Silicon Valley Suite                     0   
3  Bonsai Garden Inn in Professorville                     1   
4    Private  room - Parking 3 carport                     0   

   host_response_rate100  host_acceptance_rate100  host_is_superhost  \
0                      1                        0                  1   
1                      1                        0                  0   
2                      1                        0                  0   
3                      1                        0                  1   
4                      1                        1                  1   

   host_identity_verified  room_type_private_room  room_type_entire_home  \
0                       0                       1                      0   
1                       0                       0                      1   
2                       1                       0                      1   
3                       0                       1                      0   
4                       1                       1                      0   

   charge_extra_people_fee  cancellation_policy_flexible  \
0                        1                             0   
1                        1                             0   
2                        0                             0   
3                        1                             0   
4                        0                             0   

   cancellation_policy_moderate  cancellation_policy_strict  instant_bookable  \
0                             1                           0                 0   
1                             1                           0                 0   
2                             0                           1                 0   
3                             1                           0                 0   
4                             1                           0                 0   

   price  security_deposit  cleaning_fee  weekly_special_price  \
0   90.0               0.0          35.0                 190.0   
1   85.0             700.0         250.0                   0.0   
2   65.0            1000.0           0.0                -945.0   
3   99.0               0.0          35.0                  93.0   
4   52.0               0.0           0.0                 -29.0   

   monthly_special_price neighbourhood_cleansed  wifi  essentials  heating  \
0                 1362.0              Palo Alto     1           1        1   
1                    0.0            Santa Clara     1           0        1   
2                 -785.0          Mountain View     1           1        1   
3                 1069.0              Palo Alto     1           1        1   
4                  226.0               San Jose     1           1        1   

   smoke_alarm  hangers  shampoo  carbon_monoxide_alarm  \
0            1        1        1                      1   
1            1        1        0                      1   
2            1        1        1                      1   
3            1        1        1                      1   
4            1        1        1                      1   

   laptop-friendly_workspace  hair_dryer  kitchen  washer  dryer  iron  \
0                          1           1        1       1      1     1   
1                          1           1        1       1      1     1   
2                          1           1        0       1      1     1   
3                          1           1        1       1      1     1   
4  

# Text Normalization

Ref: [code](https://www.kaggle.com/peterwittek/wordclouds-of-titles-and-abstracts-with-n-grams)


In [ ]:
def get_bitrigrams(full_text, threshold=30):
    if isinstance(full_text, str):
        text = full_text
    else:
        text = " ".join(full_text)
    bigram_measures = BigramAssocMeasures()
    trigram_measures = TrigramAssocMeasures()
    finder = BigramCollocationFinder.from_words(text.split())
    finder.apply_freq_filter(3)
    bigrams = {
        " ".join(words): "_".join(words)
        for words in finder.above_score(bigram_measures.likelihood_ratio,
                                        threshold)
    }
    finder = TrigramCollocationFinder.from_words(text.split())
    finder.apply_freq_filter(3)
    trigrams = {
        " ".join(words): "_".join(words)
        for words in finder.above_score(trigram_measures.likelihood_ratio,
                                        threshold)
    }
    return bigrams, trigrams


def replace_bitrigrams(text, bigrams, trigrams):
    if isinstance(text, str):
        texts = [text]
    else:
        texts = text
    new_texts = []
    for t in texts:
        t_new = t
        for k, v in trigrams.items():
            t_new = t_new.replace(k, v)
        for k, v in bigrams.items():
            t_new = t_new.replace(" " + k + " ", " " + v + " ")
        new_texts.append(t_new)
    if len(new_texts) == 1:
        return new_texts[0]
    else:
        return new_texts


def process_text(text, lemmatizer, translate_table, stopwords):
    processed_text = ""
    for sentence in sent_tokenize(text):
        tagged_sentence = pos_tag(
            word_tokenize(sentence.translate(translate_table)))
        for word, tag in tagged_sentence:
            word = word.lower()
            if word not in stopwords:
                if tag[0] != 'V':
                    processed_text += lemmatizer.lemmatize(word) + " "
    return processed_text


def get_all_processed_texts(texts, lemmatizer, translate_table, stopwords):
    processed_texts = []
    for index, doc in enumerate(texts):
        processed_texts.append(
            process_text(doc, wordnet_lemmatizer, translate_table, stop))
    bigrams, trigrams = get_bitrigrams(processed_texts)
    very_processed_texts = replace_bitrigrams(processed_texts, bigrams,
                                              trigrams)
    return " ".join(very_processed_texts)

In [ ]:
# wordcloud = WordCloud(background_color="white").\
#     generate(get_all_processed_texts(records, wordnet_lemmatizer))
# plt.figure(figsize=(15, 10))
# plt.axis("off")
# plt.imshow(wordcloud)

In [ ]:
# records = airbnb['amenities'].str.lower()
# wordnet_lemmatizer = WordNetLemmatizer()
# stop = set(stopwords.words('english'))
# translate_table = dict((ord(char), " ") for char in string.punctuation)

In [ ]:
# amenitieslist =[]
# for itemlist in records:
#     amenitieslist.extend(itemlist.replace('"',' ').replace('{','').replace('}','').split(','))
# amenitiesdic = WordCloud().process_text(' '.join(amenitieslist))

# sorted([(value,key) for (key,value) in amenitiesdic.items()], reverse=True)


[(6794, 'smoke alarm'),
(6214, 'essentials shampoo'),
(6178, 'monoxide alarm'),
(6178, 'carbon monoxide'),
(6153, 'alarm carbon'),
(5998, 'laptop friendly'),
(5998, 'friendly workspace'),
(5997, 'hair dryer'),
(5768, 'washer dryer'),
(5739, 'hangers hair'),
(5537, 'dryer smoke'),
(5344, 'dryer iron'),
(5311, 'hot water'),
(5277, 'free parking'),
(5121, 'iron laptop'),
(5083, 'kitchen free'),
(4571, 'air conditioning'),
(4512, 'wifi air'),
(4285, 'heating washer'),
(4043, 'street parking'),
(4043, 'free street'),
(4040, 'fire extinguisher'),
(3913, 'extinguisher essentials'),
(3907, 'coffee maker'),
(3752, 'microwave coffee'),
(3749, 'maker refrigerator'),
(3710, 'bed linens'),
(3677, 'bedroom door'),
(3668, 'first aid'),
(3668, 'aid kit'),
(3642, 'alarm first'),
(3623, 'tv wifi'),
(3447, 'cooking basics'),
(3433, 'water bed'),
(3357, 'silverware cooking'),
(3308, 'shampoo lock'),
(3257, 'self check'),
(3201, 'oven stove'),
(3180, 'door hangers'),
(3083, 'basics oven'),
(3052, 'private entrance'),


![](http://)


In [ ]:
#loop and get the unique value for each object column
display(
    HTML(airbnb[room_colname].select_dtypes(include=['object']).T.apply(
        lambda x: x.unique(), axis=1).to_frame().to_html()))


# Clean Review

Last Review to date
